### PR TITLE
Add partial CSRF token search

### PR DESCRIPTION
**Pull request by DiogoMRSilva**
resolve zaproxy/zaproxy#6068

### CODE REVIEW

**Comment by psiinon** [comment]
Thanks for this enhancement!
But just a thought - would regex matching be more useful?

**Comment by DiogoMRSilva** [comment]
> But just a thought - would regex matching be more useful?

Not sure, I can't remember any case where I would be needed. Do you have some example where it would be useful? All the cases I can remember would be something like `.*_CSRF_.*` which is the same as using sub-string. The way I will use this is to add the keyword "CSRF" to the list.

Btw, Should I add the word "CSRF" to the list of default tokens names? I believe it would be good to make ZAP work better out of the box.

**Comment by kingthorin** [comment]
> Btw, Should I add the word "CSRF" to the list of default tokens names? I believe it would be good to make ZAP work better out of the box.

Sure.

File: zap/src/main/java/org/zaproxy/zap/extension/anticsrf/AntiCsrfParam.java [hunk 1]
```diff
@@ -43,6 +43,9 @@
     private static final String CONFIRM_REMOVE_TOKEN_KEY =
             ANTI_CSRF_BASE_KEY + ".confirmRemoveToken";
 
+    private static final String IS_PARTIAL_MATCHING_ENABLED_KEY =
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/anticsrf/AntiCsrfParam.java:46]: The `IS_` could be removed, more common without it, same for the key.

File: zap/src/main/java/org/zaproxy/zap/extension/anticsrf/AntiCsrfParam.java [hunk 2]
```diff
@@ -54,14 +57,17 @@
...
+        "__csrf_magic",
+        "CSRF"
     };
 
     private List<AntiCsrfParamToken> tokens = null;
     private List<String> enabledTokensNames = null;
 
     private boolean confirmRemoveToken = true;
 
+    private boolean isPartialMatchingEnabled = true;
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/anticsrf/AntiCsrfParam.java:69]: `partial...`

File: zap/src/main/java/org/zaproxy/zap/extension/anticsrf/AntiCsrfParam.java [hunk 3]
```diff
@@ -196,4 +203,17 @@ public void setConfirmRemoveToken(boolean confirmRemove) {
         this.confirmRemoveToken = confirmRemove;
         getConfig().setProperty(CONFIRM_REMOVE_TOKEN_KEY, confirmRemoveToken);
     }
+
+    /** Returns true if should detect CSRF variables by searching for partial matches. */
+    @ZapApiIgnore
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/anticsrf/AntiCsrfParam.java:208]: Any reason to not have this exposed in the API? (Same for the setter.)

File: zap/src/main/java/org/zaproxy/zap/extension/anticsrf/AntiCsrfParam.java [hunk 4]
```diff
@@ -196,4 +203,17 @@ public void setConfirmRemoveToken(boolean confirmRemove) {
...
+    public boolean getIsPartialMatchingEnabled() {
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/anticsrf/AntiCsrfParam.java:209]: `isPartial...`

File: zap/src/main/java/org/zaproxy/zap/extension/anticsrf/AntiCsrfParam.java [hunk 5]
```diff
@@ -196,4 +203,17 @@ public void setConfirmRemoveToken(boolean confirmRemove) {
...
+        return isPartialMatchingEnabled;
+    }
+
+    /** Define if ZAP should detect CSRF variables by searching for partial matches. */
+    @ZapApiIgnore
+    public void setIsPartialMatchingEnabled(boolean enabled) {
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/anticsrf/AntiCsrfParam.java:215]: `setPartial...`

File: zap/src/main/java/org/zaproxy/zap/extension/anticsrf/ExtensionAntiCSRF.java [hunk 6]
```diff
@@ -196,6 +197,11 @@ protected AntiCsrfParam getParam() {
         return this.getParam().getTokensNames();
     }
 
+    /** Returns true if should detect CSRF variables by searching for partial matches. */
+    public Boolean getIsPartialMatchingEnabled() {
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/anticsrf/ExtensionAntiCSRF.java:201]: Does it need to be public? The following could be called directly in the method.

File: zap/src/main/java/org/zaproxy/zap/extension/anticsrf/ExtensionAntiCSRF.java [hunk 7]
```diff
@@ -362,24 +368,14 @@ public String getTokenValue(HttpMessage tokenMsg, String tokenName) {
...
                         boolean found = false;
-                        if (attId != null) {
-                            for (String tokenName : this.getAntiCsrfTokenNames()) {
-                                if (tokenName.equalsIgnoreCase(attId)) {
-                                    list.add(new AntiCsrfToken(msg, attId, value, formIndex));
-                                    found = true;
-                                    break;
-                                }
-                            }
+                        if (attId != null && isKnownAntiCsrfToken(attId)) {
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/anticsrf/ExtensionAntiCSRF.java:371]: The `null` check could be moved to `isKnownAntiCsrfToken` as well.

**Comment by thc202** [comment]
- Could you rebase and update the tests to cover the new matching logic?
 - Could you update the help page to mention the new option? (i.e. https://github.com/zaproxy/zap-core-help/blob/a0c94fa5296b45e8450301480d8aeb4557081a3b/addOns/help/src/main/javahelp/contents/ui/dialogs/options/anticsrf.html)

**Comment by DiogoMRSilva** [comment]
>     * Could you rebase and update the tests to cover the new matching logic?
https://github.com/zaproxy/zaproxy/pull/6088/commits/2fd4a133d9b73b40fd3669e5f306da1dc6dc4b2d

>     * Could you update the help page to mention the new option? (i.e. https://github.com/zaproxy/zap-core-help/blob/a0c94fa5296b45e8450301480d8aeb4557081a3b/addOns/help/src/main/javahelp/contents/ui/dialogs/options/anticsrf.html)
https://github.com/zaproxy/zap-core-help/pull/321

I believe it is ready.

File: zap/src/main/java/org/zaproxy/zap/extension/anticsrf/AntiCsrfParam.java [hunk 8]
```diff
@@ -43,6 +43,9 @@
     private static final String CONFIRM_REMOVE_TOKEN_KEY =
             ANTI_CSRF_BASE_KEY + ".confirmRemoveToken";
 
+    private static final String PARTIAL_MATCHING_ENABLED =
+            ANTI_CSRF_BASE_KEY + ".partialMatchingEnable";
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/anticsrf/AntiCsrfParam.java:47]: Just noticed that it's missing the `d` in `Enabled`.

File: zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties [hunk 9]
```diff
@@ -11,6 +11,10 @@ acsrf.api.desc =
 acsrf.api.other.genForm = Generate a form for testing lack of anti-CSRF tokens - typically invoked via ZAP
 acsrf.api.other.genForm.param.hrefId = 
 acsrf.api.view.optionTokensNames = Lists the names of all anti-CSRF tokens
+acsrf.api.view.optionPartialMatchingEnabled = Define if ZAP should detect CSRF variables by searching for partial matches
```
Comment (thc202) [inline zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties:14]: `variables` → `tokens` (Same comment for other descriptions.)

File: zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties [hunk 10]
```diff
@@ -11,6 +11,10 @@ acsrf.api.desc =
...
+acsrf.api.action.isPartialMatchingEnabled = Returns true if should detect CSRF variables by searching for partial matches.
```
Comment (thc202) [inline zap/src/main/resources/org/zaproxy/zap/resources/Messages.properties:15]: This one should not be needed.

File: zap/src/test/java/org/zaproxy/zap/extension/anticsrf/ExtensionAntiCSRFUnitTest.java [hunk 11]
```diff
@@ -240,6 +240,49 @@ void shouldGetTokensFromInputFieldsIfIdAndNameIsAKnownTokenWithDifferentCase() {
             assertAntiCsrfToken(
                     tokens.get(1), message, KNOWN_TOKEN_2.toLowerCase(Locale.ROOT), "value2", 0);
         }
+
+        @Test
+        void
+                shouldGetTokensFromInputFieldsIfIdAndNameIncludesAKnownTokenAndPartialMatchingEnabled() {
+            given(antiCsrfParam.isPartialMatchingEnabled()).willReturn(true);
```
Comment (thc202) [inline zap/src/test/java/org/zaproxy/zap/extension/anticsrf/ExtensionAntiCSRFUnitTest.java:247]: Could you move this to after the comment? (Same comment for the following test.)

File: zap/src/main/java/org/zaproxy/zap/extension/anticsrf/AntiCsrfParam.java [hunk 12]
```diff
@@ -196,4 +203,19 @@ public void setConfirmRemoveToken(boolean confirmRemove) {
         this.confirmRemoveToken = confirmRemove;
         getConfig().setProperty(CONFIRM_REMOVE_TOKEN_KEY, confirmRemoveToken);
     }
+
+    /** Returns true if should detect CSRF variables by searching for partial matches. */
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/extension/anticsrf/AntiCsrfParam.java:207]: `variables` → `tokens` (Same comment for the following JavaDoc.)

**Comment by thc202** [comment]
Could you fix up the last two commits into the first?

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Review by kingthorin — APPROVED** [review]
Thanks :zap:

**Pull request merged**